### PR TITLE
Remove the white border in entity

### PR DESCRIPTION
**Pull request by kokosha**
The goal is to remove the white border in entity.
The alpha blending is bleeding in the border of the texture, because is sampling the transparent part of picture in weighted average, resulting in a white pixel. To correct the problem, the transparent border that is used in weighted average is recolored with color close to the opaque border.
with_border_recoloring
without_border_recoloring

The depth buffer depends on the order of drawing, if you draw a transparent pixel, you can't draw a opaque pixel before the transparent pixel, to fix the problem, we sort by the value of the distance between the camera and the entity and render from farthest to nearest distance to camera. 
depth_buffer
depth_buffer_transparent

A small fix is made in the picker, after sorting the entities, the initial entries is not the current player anymore.

### CODE REVIEW

File: korangar/src/graphics/passes/picker/entity.rs [hunk 1]
```diff
@@ -229,10 +230,10 @@ impl Prepare for PickerEntityDrawer {
 
             let mut texture_views = Vec::with_capacity_in(self.draw_count, &self.bump);
 
-            // We skip the first entity, because we don't want the player entity to show up
-            // in the picker buffer.
-            // TODO: Remove the player entity correctly.
-            for instruction in instructions.entities.iter().skip(player_size) {
+            for instruction in instructions.entities.iter() {
```
Comment (hasenbanck) [inline korangar/src/graphics/passes/picker/entity.rs:233]: I think we can use `iter().filter()` here

File: korangar/src/graphics/passes/picker/entity.rs [hunk 2]
```diff
@@ -264,10 +265,10 @@ impl Prepare for PickerEntityDrawer {
             self.instance_data_buffer.reserve(device, self.instance_data.len());
             self.bind_group = Self::create_bind_group_bindless(device, &self.bind_group_layout, &self.instance_data_buffer, &texture_views)
         } else {
-            // We skip the first entity, because we don't want the player entity to show up
-            // in the picker buffer.
-            // TODO: Remove the player entity correctly.
-            for instruction in instructions.entities.iter().skip(player_size) {
+            for instruction in instructions.entities.iter() {
```
Comment (hasenbanck) [inline korangar/src/graphics/passes/picker/entity.rs:268]: I think we can use iter().filter() here too.

File: korangar/src/world/map/mod.rs [hunk 3]
```diff
@@ -265,10 +265,27 @@ impl Map {
...
     #[cfg_attr(feature = "debug", korangar_debug::profile)]
     pub fn render_entities(&self, instructions: &mut Vec<EntityInstruction>, entities: &[Entity], camera: &dyn Camera, include_self: bool) {
-        entities
-            .iter()
-            .skip(!include_self as usize)
-            .for_each(|entity| entity.render(instructions, camera));
+        fn get_distance(camera: &dyn Camera, entity: &Entity) -> f32 {
+            camera.distance_to(entity.get_position())
+        }
+        let mut entity_order: Vec<usize> = (0..entities.len()).collect();
```
Comment (hasenbanck) [inline korangar/src/world/map/mod.rs:271]: We should try to not allocate every frame inside the hot loop.

File: korangar/src/world/map/mod.rs [hunk 4]
```diff
@@ -265,10 +265,27 @@ impl Map {
...
+
+        entity_order.sort_by(|index_a, index_b| {
+            let dist_a = get_distance(camera, &entities[*index_a]);
+            let dist_b = get_distance(camera, &entities[*index_b]);
+            dist_b.partial_cmp(&dist_a).unwrap()
+        });
+
+        entity_order.iter().for_each(|&index| {
+            let is_self = index == 0;
```
Comment (hasenbanck) [inline korangar/src/world/map/mod.rs:280]: "self" has a pretty well defined meaning in Rust. We should better use "is_player_entity" or somthing similar.

**Comment by kokosha** [comment]
The new way is changing the alpha from (0.0;1.0] into 1.0, when the entity is opaque. When the entity is transparent, we need to sort to treat the ordering, but even sorting won't solve it all, as the first transparent pixel will overwrite the depth buffer, you will need new techniques such as depth peeling or other order-independent transparency method and is process intensive to fix transparency. Another problem is the way that the entities is structured in the moment doesn't allow easy sorting without affecting hot loop.

**Comment by kokosha** [comment]
[156](https://github.com/vE5li/korangar/pull/156),  [157](https://github.com/vE5li/korangar/pull/157), fixed the problem

**Pull request closed without merging**